pravega/src/main/java/akka/stream/alpakka/pravega/javadsl/Pravega.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.pravega.javadsl;
+
+import akka.Done;
+import akka.NotUsed;
+import akka.stream.alpakka.pravega.WriterSettings;
+import akka.stream.javadsl.Flow;
+import akka.stream.javadsl.Keep;
+import akka.stream.javadsl.Sink;
+import akka.stream.javadsl.Source;
+
+import io.pravega.client.stream.ReaderConfig;
+import scala.compat.java8.FutureConverters;
+import static scala.compat.java8.OptionConverters.toScala;
+
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+import akka.stream.alpakka.pravega.impl.PravegaFlow;
+import akka.stream.alpakka.pravega.impl.PravegaSource;
+
+import akka.stream.alpakka.pravega.ReaderSettings;
+
+import io.pravega.client.ClientConfig;
+import io.pravega.client.stream.EventStreamWriter;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.Serializer;
+
+public class Pravega {
+
+  /**
+   * Messages are read from a Pravega stream.
+   *
+   * <p>Materialized value is a future which completes to Done as soon as Pravega reader in open.
+   */
+  public static <V> Source<V, CompletionStage<Done>> source(
+      String scope, String topic, ReaderSettings<V> settings) {
+    return Source.fromGraph(new PravegaSource<V>(scope, topic, settings))
+        .mapMaterializedValue(FutureConverters::<Done>toJava);
+  }
+
+  /** Incoming messages are written to Pravega stream and emitted unchanged. */
+  public static <V> Flow<V, V, NotUsed> flow(
+      String scope,
+      String streamName,
+      Optional<String> routingKey,
```

Comment:
According to the Pravega docs, routing keys are analagous to message keys in Kafka.  They're required to make certain that messages produced to the same key will be written to and read from in the same order.  Instead of a routing key as a parameter here we should create types that the user can wrap their messages in that include an optional key.

See Alpakka Kafka for reference
* https://github.com/akka/alpakka-kafka/blob/master/core/src/main/scala/akka/kafka/ProducerMessage.scala
* https://github.com/akka/alpakka-kafka/blob/master/core/src/main/scala/akka/kafka/ConsumerMessage.scala

I don't think we need the same level of complexity, but we can create a simple Producer and Consumer message type with factory methods to include or not include a routing key.  They can serve as I/O of the Pravega source, flow, and sink stages.

http://pravega.io/docs/latest/key-features/#pravega-streams-are-ordered

Reply:
😊 stupid I am ... sorry.

Reply:
No you're not! Simple misunderstanding. :)

Reply:
I could also provide a function per sink to extract the key, might be far easier to use also.
I cannot really see a use case where the key extraction with change between 2 messages.
WDYT ?
Pushed this option, I also remove the scala2.11 cross compilation (because of the extraction function :-/ )

Reply:
Agreed. I think a UDF should be good enough.

I'm not too concerned about supporting 2.11, but I'm not sure I understand the problem.  Are you referring to SAM type conversion for the lambda in the Java DSL?

Reply:
@cheleb & @seglo -- In Pravega, as stream segments split and merge, different keys get shuffled between readers and their segments, and keys are lost on read. If users don't embed the key in their event or message, then the key cannot be recovered on read.

Reply:
Got it. Yes, we'll only expose a way for the user to define the routing key on the write side (for akka streams Flows and Sinks).

Reply:
> Agreed. I think a UDF should be good enough.
> 
> I'm not too concerned about supporting 2.11, but I'm not sure I understand the problem. Are you referring to SAM type conversion for the lambda in the Java DSL?
Yes, scala.compat.java8 is in pain in idea ..., but I can make it, AFAIR I've encoutered the same issue with Geode connector.

Reply:
I think this thread is resolved.  Please re-open if you feel there's more to discuss.